contracts/src/libraries/token/erc677.cairo

```diff
@@ -0,0 +1,40 @@
+use starknet::ContractAddress;
+
+#[abi]
+trait IERC677Receiver {
+    fn on_token_transfer(sender: ContractAddress, value: u256, data: Array<felt252>);
+    // implements EIP-165, where function selectors are defined by Ethereum ABI using the ethereum function signatures
+    fn supports_interface(interface_id: felt252) -> bool;
+}
+
+mod ERC677 {
+    use starknet::ContractAddress;
+    use chainlink::libraries::token::erc20::ERC20;
+    use array::ArrayTrait;
+    use array::SpanTrait;
+    use clone::Clone;
+    use array::ArrayTCloneImpl;
+
+    use super::IERC677ReceiverDispatcher;
+    use super::IERC677ReceiverDispatcherTrait;
+
+    const IERC677_RECEIVER_ID: felt252 = 0xa53f2491;
+
+    #[event]
+    fn Transfer(from: ContractAddress, to: ContractAddress, value: u256, data: Array<felt252>) {}
+
+    fn transfer_and_call(to: ContractAddress, value: u256, data: Array<felt252>) -> bool {
+        let sender = starknet::info::get_caller_address();
+
+        ERC20::transfer(to, value);
+        Transfer(sender, to, value, data.clone());
```

Comment:
had to use clone here because events do not support references. I thought about using @ (snapshot) but I'd have to implement Serde on Array<felt252> type which involves desnapping, but desnapping is only valid on Copy-able but Array isn't copyable

Reply:
I think a clone here is fine! You need a copy of the data for both calls